includes/abstracts/abstract-wc-product.php

```diff
@@ -1425,54 +1406,93 @@ public function delete() {
-		update_post_meta( $id, '_length', $this->get_length() );
-		update_post_meta( $id, '_width', $this->get_width() );
-		update_post_meta( $id, '_height', $this->get_height() );
-		update_post_meta( $id, '_upsell_ids', $this->get_upsell_ids() );
-		update_post_meta( $id, '_crosssell_ids', $this->get_cross_sell_ids() );
-		update_post_meta( $id, '_purchase_note', $this->get_purchase_note() );
-		update_post_meta( $id, '_default_attributes', $this->get_default_attributes() );
-		update_post_meta( $id, '_virtual', $this->get_virtual() ? 'yes' : 'no' );
-		update_post_meta( $id, '_downloadable', $this->get_downloadable() ? 'yes' : 'no' );
-		update_post_meta( $id, '_product_image_gallery', implode( ',', $this->get_gallery_attachment_ids() ) );
-		update_post_meta( $id, '_download_limit', $this->get_download_limit() );
-		update_post_meta( $id, '_download_expiry', $this->get_download_expiry() );
-
-		if ( update_post_meta( $id, '_featured', $this->get_featured() ) ) {
-			delete_transient( 'wc_featured_products' );
-		}
+		$updated_props     = array();
+		$meta_key_to_props = array(
+			'_visibility'            => 'catalog_visibility',
+			'_sku'                   => 'sku',
+			'_regular_price'         => 'regular_price',
+			'_sale_price'            => 'sale_price',
+			'_sale_price_dates_from' => 'date_on_sale_from',
+			'_sale_price_dates_to'   => 'date_on_sale_to',
+			'total_sales'            => 'total_sales',
+			'_tax_status'            => 'tax_status',
+			'_tax_class'             => 'tax_class',
+			'_manage_stock'          => 'manage_stock',
+			'_backorders'            => 'backorders',
+			'_sold_individually'     => 'sold_individually',
+			'_weight'                => 'weight',
+			'_length'                => 'length',
+			'_width'                 => 'width',
+			'_height'                => 'height',
+			'_upsell_ids'            => 'upsell_ids',
+			'_crosssell_ids'         => 'cross_sell_ids',
+			'_purchase_note'         => 'purchase_note',
+			'_default_attributes'    => 'default_attributes',
+			'_virtual'               => 'virtual',
+			'_downloadable'          => 'downloadable',
+			'_product_image_gallery' => 'gallery_attachment_ids',
+			'_download_limit'        => 'download_limit',
+			'_download_expiry'       => 'download_expiry',
+			'_featured'              => 'featured',
+			'_thumbnail_id'          => 'thumbnail_id',
+			'_downloadable_files'    => 'downloads',
+			'_stock'                 => 'stock_quantity',
+			'_stock_status'          => 'stock_status',
+		);
 
-		if ( ! empty( $this->get_thumbnail_id() ) ) {
-			set_post_thumbnail( $id, $this->get_thumbnail_id() );
-		} else {
-			delete_post_meta( $id, '_thumbnail_id' );
+		foreach ( $meta_key_to_props as $meta_key => $prop ) {
+			$value = $this->data[ $prop ];
+			// @todo this is where state should be checked?
+			switch ( $prop ) {
+				case 'virtual' :
+				case 'downloadable' :
+					$updated = update_post_meta( $this->get_id(), $meta_key, $value ? 'yes' : 'no' );
```

Comment:
We should introduce a `wc_bool_to_string` counterpart to `wc_string_to_bool`. It could simplify this and some other lines.